### ISSUE TITLE
[DATA DEFECT] Class imbalance in Sentiment140 sample - 569 negative vs 431 positive (0.757 ratio)

### ISSUE THREAD

**Issue by Ahmad-AbuGhazaleh** [Ahmad-AbuGhazaleh/DS-AI-SQA-Project#5]
## Bug Description
Random sampling without stratification caused class imbalance in the Sentiment140 dataset sample. The negative class has 32% more samples than positive class.

## Current State
- **Total samples:** 1,000 tweets
- **Negative (0):** 569 tweets (56.9%)
- **Positive (4):** 431 tweets (43.1%)
- **Balance ratio:** 0.757 (should be ~1.0)
- **Imbalance:** 24.3% deviation from perfect balance

## Expected Behavior
According to ML best practices and SQA Plan requirements:
- Classes should be balanced (50/50 split)
- Balance ratio should be 0.95 - 1.05 (within 5% tolerance)
- Stratified sampling should be used to maintain original class distribution

## Impact
- **Severity:** Medium
- **Type:** Data Defect
- Model will be biased toward predicting negative sentiment
- Accuracy metric will be misleading (baseline accuracy = 56.9%)
- Precision/Recall metrics will be skewed
- Violates fair classification principles

## Root Cause
```python
# Current implementation (WRONG)
df_sample = df.sample(n=sample_size)  # No stratification
```

Random sampling doesn't guarantee class balance, especially with small sample sizes.

## Proposed Solution
Use stratified sampling to maintain 50/50 class balance:
```python
# Proposed implementation
from sklearn.model_selection import train_test_split

# Stratified sampling
df_sample, _ = train_test_split(
    df, 
    train_size=sample_size, 
    stratify=df['target'],
    random_state=42
)
```

This ensures exactly 500 positive and 500 negative samples.

## Verification Criteria
- ✅ Positive samples: 500 ± 5
- ✅ Negative samples: 500 ± 5
- ✅ Balance ratio: 0.95 - 1.05
- ✅ Both classes equally represented

---
**Reported by:** Ahmad Abu Ghazaleh  
**Student ID:** 202210221  
**Date:** December 27, 2025
